core/src/main/scala/akka/persistence/r2dbc/query/scaladsl/R2dbcReadJournal.scala

```diff
@@ -197,6 +231,71 @@ final class R2dbcReadJournal(system: ExtendedActorSystem, config: Config, cfgPat
       dbSource
   }
 
+  /**
+   * Same as `eventsBySlices` but with the purpose to use snapshots as starting points and thereby reducing number of
+   * events that have to be loaded. This can be useful if the consumer start from zero without any previously processed
+   * offset or if it has been disconnected for a long while and its offset is far behind.
+   *
+   * First it loads all snapshots with timestamps greater than or equal to the offset timestamp. There is at most one
+   * snapshot per persistenceId. The snapshots are transformed to events with the given `transformSnapshot` function.
+   *
+   * After emitting the snapshot events the ordinary events with sequence numbers after the snapshots are emitted.
+   */
+  def eventsBySlicesStartingFromSnapshots[Snapshot, Event](
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset,
+      transformSnapshot: Snapshot => Event): Source[EventEnvelope[Event], NotUsed] = {
+    // FIXME DRY with above eventsBySlices
+
+    val snapshotSource =
+      snapshotsBySlice[Snapshot, Event](transformSnapshot)
+        .currentBySlices("snapshotsBySlices", entityType, minSlice, maxSlice, offset)
+
+    val dbSource =
+      bySlice[Event].liveBySlices(
+        "eventsBySlices",
+        entityType,
+        minSlice,
+        maxSlice,
+        offset, // FIXME adjust based on snapshots
+        snapshotOffsets = Map.empty // FIXME collect offsets corresponding to snapshots
```

Comment:
I want to collect these offsets from the emitted snapshot events from snapshotSource before starting the dbSource. Can that be done with existing Akka Streams operators or do I have to implement a graph stage?

Reply:
thanks, I'll try

Reply:
I guess if we anyway side effect it out, doing that directly from `onComplete` in `statefulMap` and then returning `None` is just as good and less messy.

Reply:
The problem is that we return a Source and we are not materializing it ourselves, so it may be materialized more than once and then I can't side effect via a single Promise.

Reply:
You can stuff it all inside a `Source.setup` to tie it to materialization

Reply:
implemented in https://github.com/akka/akka-persistence-r2dbc/pull/410/commits/563a97cfd9a5b7d7a63b4692c8492fbaa176f282

Reply:
ah, I missed the existence of Source.setup. Don't you think it's anyway more clean with a custom graph stage. It wasn't too involved.

Reply:
Yep, looks custom stage like a good path to me. 👍